spaceship/lib/spaceship/connect_api/models/device.rb

```diff
@@ -54,8 +54,26 @@ def self.all(client: nil, filter: {}, includes: nil, limit: nil, sort: nil)
         return resps.flat_map(&:to_models)
       end
 
+      # @param client [ConnectAPI] ConnectAPI client.
+      # @param platform [String] The platform of the device.
+      # @param include_disabled [Bool] Whether to include disable devices. false by default.
+      # @return (Device) Find a device based on the UDID of the device. nil if no device was found.
+      def self.find_by_udid(device_udid, client: nil, platform: nil, include_disabled: false)
+        self.all(client: client).find do |device|
+          device.udid.casecmp(device_udid) == 0 && (include_disabled ? true : device.enabled?)
+        end
+      end
+
       def self.create(client: nil, name: nil, platform: nil, udid: nil)
         client ||= Spaceship::ConnectAPI
+
+        # Find the device by UDID and return it in case it already exists
+        existing = self.find_by_udid(udid, client: client, platform: platform)
+        if existing
+          UI.verbose("UDID '#{udid}' already exists - Skipping its creation...")
+          return existing
+        end
+
         resp = client.post_device(name: name, platform: platform, udid: udid)
```

Comment:
So I don't know a lot about fastlane's inner workings or best practices, but could another approach be to check/catch if this POST request returns a 409 Conflict and simply ignore it (well, maybe log a message that the device already exists like you're doing above), since it would mean the device already exists. Just thinking that might be a simpler solution and shave off some time from the register_device operation.

Reply:
I copied the same approach used in the old (now deprecated) method that did the same thing using the iTunes undocumented APIs: instead of recovering from an error after sending the request, it fails the request early before sending it to the server. The server could still return a 409 if a) the local check fails or b) if the request is invalid for other reasons, but I think that's okay